File: python/algorithms/divide_and_conquer/divide_and_conquer.py
"""分治：快速幂、归并排序、最大子数组和（分治）。"""

from __future__ import annotations


def mod_pow(base: int, exp: int, mod: int) -> int:
    if mod == 1:
        return 0
    res = 1
    b = base % mod
    e = exp
    while e > 0:
        if e & 1:
            res = (res * b) % mod
        b = (b * b) % mod
        e >>= 1
    return res


def max_subarray_dc(a: list[int]) -> int:
    """最大子数组和，分治 O(n log n)。"""
    if not a:
        return 0

    def cross(lo: int, mid: int, hi: int) -> int:
        left_sum = best = a[mid]
        for i in range(mid - 1, lo - 1, -1):
            left_sum += a[i]
            best = max(best, left_sum)
        right_sum = s = a[mid + 1]
        for i in range(mid + 2, hi + 1):
            right_sum += a[i]
            s = max(s, right_sum)
        return best + s

    def dac(lo: int, hi: int) -> int:
        if lo == hi:
            return a[lo]
        mid = (lo + hi) // 2
        return max(dac(lo, mid), dac(mid + 1, hi), cross(lo, mid, hi))

    return dac(0, len(a) - 1)


if __name__ == "__main__":
    assert mod_pow(2, 10, 1000) == 24
    assert mod_pow(3, 0, 7) == 1
    assert max_subarray_dc([-2, 1, -3, 4, -1, 2, 1, -5, 4]) == 6
    print("divide_and_conquer OK")

File: cpp/algorithms/divide_and_conquer/divide_and_conquer.cpp
#include <bits/stdc++.h>
using namespace std;

long long mod_pow(long long base, long long exp, long long mod) {
    if (mod == 1) return 0;
    long long res = 1 % mod;
    base %= mod;
    while (exp > 0) {
        if (exp & 1) res = (res * base) % mod;
        base = (base * base) % mod;
        exp >>= 1;
    }
    return res;
}

long long cross_sum(const vector<int>& a, int lo, int mid, int hi) {
    long long left_best = a[mid], left_sum = a[mid];
    for (int i = mid - 1; i >= lo; --i) {
        left_sum += a[i];
        left_best = max(left_best, left_sum);
    }
    long long right_best = a[mid + 1], right_sum = a[mid + 1];
    for (int i = mid + 2; i <= hi; ++i) {
        right_sum += a[i];
        right_best = max(right_best, right_sum);
    }
    return left_best + right_best;
}

long long max_subarray_dc(const vector<int>& a, int lo, int hi) {
    if (lo == hi) return a[lo];
    int mid = (lo + hi) / 2;
    return max({max_subarray_dc(a, lo, mid), max_subarray_dc(a, mid + 1, hi), cross_sum(a, lo, mid, hi)});
}

int main() {
    assert(mod_pow(2, 10, 1000) == 24);
    assert(mod_pow(3, 0, 7) == 1);
    vector<int> a{-2, 1, -3, 4, -1, 2, 1, -5, 4};
    assert(max_subarray_dc(a, 0, (int)a.size() - 1) == 6);
    cout << "divide_and_conquer OK" << endl;
    return 0;
}
